cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -218,10 +230,94 @@ func newMachineController(
 	}
 
 	return &machineController{
-		clusterInformerFactory: clusterInformerFactory,
-		kubeInformerFactory:    kubeInformerFactory,
-		machineInformer:        machineInformer,
-		machineSetInformer:     machineSetInformer,
-		nodeInformer:           nodeInformer,
+		clusterClientset:          clusterclient,
+		clusterInformerFactory:    clusterInformerFactory,
+		kubeInformerFactory:       kubeInformerFactory,
+		machineDeploymentInformer: machineDeploymentInformer,
+		machineInformer:           machineInformer,
+		machineSetInformer:        machineSetInformer,
+		nodeInformer:              nodeInformer,
 	}, nil
 }
+
+func (c *machineController) machineSetNodeNames(machineSet *v1alpha1.MachineSet) ([]string, error) {
+	machines, err := c.MachinesInMachineSet(machineSet)
+	if err != nil {
+		return nil, fmt.Errorf("error listing machines: %v", err)
+	}
+
+	var nodes []string
+
+	for _, machine := range machines {
+		if machine.Status.NodeRef == nil {
+			glog.V(4).Infof("Status.NodeRef of machine %q is currently nil", machine.Name)
+			continue
+		}
+		if machine.Status.NodeRef.Kind != "Node" {
+			glog.Errorf("Status.NodeRef of machine %q does not reference a node (rather %q)", machine.Name, machine.Status.NodeRef.Kind)
+			continue
+		}
+
+		node, err := c.findNodeByNodeName(machine.Status.NodeRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("unknown node %q", machine.Status.NodeRef.Name)
+		}
+
+		if node != nil {
+			nodes = append(nodes, node.Spec.ProviderID)
+		}
+	}
+
+	glog.V(4).Infof("nodegroup %s has nodes %v", machineSet.Name, nodes)
+
+	return nodes, nil
+}
+
+func (c *machineController) nodeGroups() []cloudprovider.NodeGroup {
+	machineSets, err := c.machineSetInformer.Lister().MachineSets(metav1.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		return nil
+	}
+
+	var nodegroups []cloudprovider.NodeGroup
+
+	for _, machineSet := range machineSets {
```

Comment:
There might be multiple machineSets backed by the same machineDeployment, hence this would incorrectly return multiple `nodeGroups` for the same backing `deployment`, right?

Reply:
Yes, as discussed f2f will scan for MachineDepoyments independently and only pickup machineSets that have no machineDeployment owner reference.

Reply:
Done in https://github.com/openshift/kubernetes-autoscaler/pull/26/commits/e7a117ce4cbee3bdd6ea6de1a653d7f28414a980

Reply:
Correction; done in https://github.com/openshift/kubernetes-autoscaler/pull/26/commits/48177bc9f5ca4db4250695aa8995219a64376915

Reply:
I squashed all the WIP commits; this commit has gone.